src/iop/crop.c

```diff
@@ -0,0 +1,1801 @@
+
+  dt_control_queue_redraw_center();
+}
+
+static gint _aspect_ratio_cmp(const dt_iop_crop_aspect_t *a, const dt_iop_crop_aspect_t *b)
+{
+  // want most square at the end, and the most non-square at the beginning
+
+  if((a->d == 0 || a->d == 1) && a->n == 0) return -1;
+
+  const float ad = MAX(a->d, a->n);
+  const float an = MIN(a->d, a->n);
+  const float bd = MAX(b->d, b->n);
+  const float bn = MIN(b->d, b->n);
+  const float aratio = ad / an;
+  const float bratio = bd / bn;
+
+  if(aratio < bratio) return -1;
+
+  const float prec = 0.0003f;
+  if(fabsf(aratio - bratio) < prec) return 0;
+
+  return 1;
+}
+
+static gchar *_aspect_format(gchar *original, int adim, int bdim)
+{
+  // Special ratios:  freehand, original image
+  if(bdim == 0) return g_strdup(original);
+
+  return g_strdup_printf("%s  %4.2f", original, (float)adim / (float)bdim);
+}
+
+static void _event_margins_button_changed(GtkDarktableToggleButton *widget, gpointer user_data)
+{
+  dt_iop_module_t *self = (dt_iop_module_t *)user_data;
+  dt_iop_crop_gui_data_t *g = (dt_iop_crop_gui_data_t *)self->gui_data;
+  const gboolean active = gtk_toggle_button_get_active(GTK_TOGGLE_BUTTON(g->margins_toggle));
+  dtgtk_expander_set_expanded(DTGTK_EXPANDER(g->margins_expander), active);
+  dtgtk_togglebutton_set_paint(DTGTK_TOGGLEBUTTON(g->margins_toggle), dtgtk_cairo_paint_solid_arrow,
+                               CPF_STYLE_BOX | (active ? CPF_DIRECTION_DOWN : CPF_DIRECTION_LEFT), NULL);
+  g->expand_margins = active;
+  dt_conf_set_bool("plugins/darkroom/crop/expand_margins", active);
+}
+
+static void _event_margins_expander_click(GtkWidget *widget, GdkEventButton *e, gpointer user_data)
+{
+  if(e->type == GDK_2BUTTON_PRESS || e->type == GDK_3BUTTON_PRESS) return;
+
+  dt_iop_module_t *self = (dt_iop_module_t *)user_data;
+  dt_iop_crop_gui_data_t *g = (dt_iop_crop_gui_data_t *)self->gui_data;
+
+  gtk_toggle_button_set_active(GTK_TOGGLE_BUTTON(g->margins_toggle),
+                               !gtk_toggle_button_get_active(GTK_TOGGLE_BUTTON(g->margins_toggle)));
+}
+
+void gui_init(struct dt_iop_module_t *self)
+{
+  dt_iop_crop_gui_data_t *g = IOP_GUI_ALLOC(crop);
+
+  g->aspect_list = NULL;
+  g->clip_x = g->clip_y = g->handle_x = g->handle_y = 0.0;
+  g->clip_w = g->clip_h = 1.0;
+  g->clip_max_x = g->clip_max_y = 0.0;
+  g->clip_max_w = g->clip_max_h = 1.0;
+  g->clip_max_pipe_hash = 0;
+  g->cropping = 0;
+  g->shift_hold = FALSE;
+  g->ctrl_hold = FALSE;
+  g->old_width = g->old_height = -1;
+
+  g->expand_margins = dt_conf_get_bool("plugins/darkroom/crop/expand_margins");
+
+  GtkWidget *box_enabled = gtk_box_new(GTK_ORIENTATION_VERTICAL, DT_BAUHAUS_SPACE);
+
+  GtkWidget *lb = dt_ui_section_label_new(_("crop settings"));
+  GtkStyleContext *context = gtk_widget_get_style_context(lb);
+  gtk_style_context_add_class(context, "section_label_top");
+  gtk_box_pack_start(GTK_BOX(box_enabled), lb, TRUE, TRUE, 0);
+
+  dt_iop_crop_aspect_t aspects[] = {
```

Comment:
I would move that into a new header file `src/common/crop.h` and share this with current clipping module.

There is also some duplicated routine in there that could probably be moved there too like  _aspect_ratio_cmp() and passing numerator & denominator instead of a & b of course. And I'm sure there is many others.

Reply:
_float_to_fract() is a good candidate too.

Reply:
well, that was my first goal, but after some though, I've prefered not to do so : see #8834 : I would love to see clipping deprecated (or completely rewritten) soon (not for 3.6 don't worry) so better not rely the 2 modules together... That will also give us more freedom to change things here if needed (because changing anything in clipping is.... )

Reply:
I see, makes sense yes.